File: arrays_and_hashing/valid_sudoku.rb
# frozen_string_literal: true

# Time complexity: O(n)
# Space complexity: O(n)
def valid_sudoku?(board)
  combinations = Hash.new { |h, k| h[k] = Array.new(9) }

  9.times do |r|
    row = "row-#{r}"
    9.times do |c|
      next if board[r][c] == '.'

      val = board[r][c].to_i
      col = "col-#{c}"
      box = "box-#{(c / 3) + (r / 3) * 3}"

      return false if combinations[row][val] || combinations[box][val] || combinations[col][val]

      combinations[row][val] = true
      combinations[col][val] = true
      combinations[box][val] = true
    end
  end

  true
end

File: arrays_and_hashing/valid_sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        populated = {
            'row' : [[False] * 9 for _ in range(9)],
            'col' : [[False] * 9 for _ in range(9)],
            'box' : [[False] * 9 for _ in range(9)]
        }

        for r in range(len(board)):
            for c in range(len(board[r])):
                if board[r][c] == '.':
                    continue
                
                val = int(board[r][c]) - 1
                b = ((r // 3) * 3) + ((c // 3) * 1)
                
                if populated['row'][r][val] or populated['col'][c][val] or populated['box'][b][val]:
                    return False
                
                populated['row'][r][val] = True
                populated['col'][c][val] = True
                populated['box'][b][val] = True

        return True